Wait for sent statuses before sending new ones

estan opened the pull request:
Both as a workaround for https://gitlab.com/gitlab-org/gitlab/-/issues/13134 but also to generally preclude the possibility of reports being sent out of order.

@p12tic @tardyp @g-arjones @kalvdans

Closes #6563

## Contributor Checklist:

* [x] I have updated the unit tests
* [x] I have created a file in the `newsfragments` directory (and read the `README.txt` in that directory)
* [ ] ~~I have updated the appropriate documentation~~

p12tic reviewed (changes requested):
Thanks for the PR. I think it makes sense in principle. Given this is more like a workaround for Gitlab, we need a way to turn this off. Something like `serialize_send_message` boolean argument for ReporterBase. Then Gitlab reporters could pass `serialize_send_message=True` to `super().__init__`.

Additionally, it seems that only `sendMessage` call needs protection. Can we add it only there?

Something like: 

```
if self.serialize_send_message:
    while self._event_queue:
        yield self._event_queue.popleft()
    event = defer.Deferred()
    self._event_queue.append(event)
    yield self.sendMessage(reports)
    event.callback(None)
else:
    yield self.sendMessage(reports)
```

Also rename `_event_queue` to something like `_pending_send_message_calls` and then `event` to just `d`.

Finally, add a newsfragment and this PR will be good to go. Thanks!

p12tic commented:
> I thought of this, but is there not a (small) possibility that, while _got_event is yielding on a g.generate call on line 101, another event arrives and is processed in full (including the sendMessage call), and thus the order of the sendMessage calls would no longer be guaranteed locked to the order in which the corresponding events arrived on the master event queue? This is the reason I chose to treat the generation of the reports and the sending of them using sendMessage as one unit.

@estan Indeed, you're correct, thanks. 

I propose to move the current code of `_got_event` into e.g. `_got_event_impl`, so that we only need to check `self.serialize_send_message` once.

estan commented:
The more I think about it, is it not more robust to always do this, i.e. not make it an option?

Don't we always want messages to be sent in the order in which the corresponding events arrive on the event queue?

For example GitLab (unrelated to the race condition bug) will respond with a 400 error if a nonsensical sequence of status updates arrive (for example "pending" after "running"), which could theoretically happen with the code as it is now, since _got_event yields on g.generate before sending the message, and I guess the same could be true of other (current or future) reporting destinations (?)

p12tic commented:
> Don't we always want messages to be sent in the order in which the corresponding events arrive on the event queue?

You're correct again. Could you add a comment to the code saying that since `g.generate()` is asynchronous the `sendMessage()` calls may arrive out of sequence, therefore we need to always serialize whole function?

Thanks a lot.

p12tic reviewed (changes requested):
We need a release note (see newsfragments/README.md file) and a comment explaining why the current approach is needed for all reporters, not just GitLab.

Otherwise, this PR is good to go.

p12tic commented on master/buildbot/reporters/base.py:
```diff
@@ -81,6 +88,12 @@ def _does_generator_want_key(self, generator, key):
+        while self._event_queue:
+            yield self._event_queue.popleft()
+
+        event = defer.Deferred()
```
Please rename `_event_queue` to `_pending_got_event_calls` and `event` to just `d`.

estan commented:
@p12tic Thanks for the review! I pushed up some changes. I added some comments, but feel free to reword them if you want. Also added a couple of tests.

estan commented:
Hold on, I went back to testing this on our server, to make double-sure, and it seems despite my intentions the patch is not doing what it should be doing.

To test, I surrounded the `res = yield self.createStatus(...)` line in the `sendMessage` method in gitlab.py with `log.msg(f"==== {sha} {state}")` and `log.msg(f"==== {res.code}")`, and when pushing to one of our projects I am seeing multiple printouts of the first log.msg, indicating messages are being sent out before pending ones have been fully handled.

Will have to investigate...

estan commented:
Yes, it was needlessly complex with a local queue and didn't work properly due to a thinko.

It's enough to track just the previous call to _got_event and wait on that. I pushed what I think is a better approach.

estan commented:
Hm, not sure how the coverage of buildbot_worker/tunnel.py was lowered by my changes.

estan commented:
I think it may just be something which is flaky, maybe rerun the workflow and see if it goes green.

p12tic reviewed (approved):
Great PR, with tests and everything.

p12tic commented:
I combined the commits into single one, as most of the commits were fixing issues in the earlier ones.

(Github UI shows me as the author of the commit. This is artifact of Github caching as I've kept the original attribution on the commit).

g-arjones commented:
@p12tic Would you mind releasing buildbot with this patch? The latest release was back in March, there are quite a few fixes in master already.

g-arjones commented:
@estan Thank you for the effort.

p12tic commented:
> Would you mind releasing buildbot with this patch?

I will look into that over the following days.
